### PR TITLE
Integrated ParticleNet

### PR DESCRIPTION
**Pull request by ccmfb**

### CODE REVIEW

File: configs/model/particle_net.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+      [16, [64, 64, 64]],
+      [16, [128, 128, 128]],
+      [16, [256, 256, 256]]
+    ]
+    dense_params: [256, 256]
+    conv_pooling: 'average'
+    dropout_rate: 0.1
+    seq_cutoff_len: 125
+  decoder:
+    output_classes: 2
```
Comment (ofivite) [inline configs/model/particle_net.yaml]: for the sake of consistent naming with other models, it would be better to rename `output_classes` -> `n_outputs `

File: models/particle_net.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+                if dropout_rate is not None and dropout_rate > 0:
+                    self.dense_dropout.append(
+                        Dropout(dropout_rate)
+                    )
+            
+            self.out = Dense(self.num_classes, activation='softmax')
+
+    @tf.function
+    def call(self, input):
+        masking = True
```
Comment (ofivite) [inline models/particle_net.py]: Move `masking` to `particle_net.yaml` under `kwargs.encoder` so that it can be configured from there

File: models/particle_net.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+        features = input[0][:,:self.num_points,:].to_tensor() # (batch_size, particles, features), here: (128, 125, 22)
+        # print(f'FEATURES - type: {type(features)}, shape: {features.shape}')
+
+        # Converting from (r, theta)-space to (eta, phi)-space
```
Comment (ofivite) [inline models/particle_net.py]: I propose to generalise the definitions of `points` to a user-defined set of features. Namely, to add parameters `spatial_features: ["r", "theta"]`  and `from_polar: True` to `kwargs.encoder` in the yaml cfg. 

Then, analogously to the Transformer/Taco cases, a mapping `feature_name_to_idx` from feature names to their indices can be passed to the ParticleNet class, so that internally one will be able to operate with names instead of hardcoded indices.

After that, if `from_polar==False`, `points` are just extracted from the inputs by fetching the indices corresponding to `spatial_features` from `feature_name_to_idx` and then doing something like `tf.gather`.

Otherwise (`from_polar==True`) asserting if `spatial_features != ["r", "theta"]`, and then doing the current L51-56 with 0th index -> `self.feature_name_to_idx['r']`, 1th index -> `self.feature_name_to_idx['theta']`

File: models/particle_net.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        phi = phi[:, :, tf.newaxis]
+
+        points = tf.concat([eta, phi], -1) # (128, 125, 2)
+
+        if masking:
+            reduced_fts = tf.expand_dims(tf.reduce_sum(features,-1),-1)  # From (128, 125, 22) to (128, 125, 1)
+            mask = tf.cast(tf.not_equal(reduced_fts, 0), dtype='float32')
+            coord_shift = tf.multiply(1e9, tf.cast(tf.equal(mask, 0), dtype='float32'))
+        
+        fts = tf.squeeze(self.batch_norm(tf.expand_dims(features, axis=2)), axis=2)
```
Comment (ofivite) [inline models/particle_net.py:91]: I am not sure that I understand what this batch norm thing does, is it actually doing layer normalisation? so like normalising only on per token basis?

File: models/particle_net.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+        for layer_idx, layer_param in enumerate(self.conv_params):
+            if masking:
+                pts = tf.add(coord_shift, points) if layer_idx == 0 else tf.add(coord_shift, fts)
+            else:
+                pts = points
```
Comment (ofivite) [inline models/particle_net.py]: It looks to me that in case of no masking, `pts` will always stay fixed to the eta/phi coordinates, while in ParticleNet it is dynamic, i.e. that `pts = fts` (except for the `layer_idx == 0`) in the loop . Or am I mistaken?

File: models/particle_net.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        eta = features[:,:,0] * tf.math.cos(features[:,:,1]) # (128, 125)
+        phi = features[:,:,0] * tf.math.sin(features[:,:,1])
+        eta = eta[:, :, tf.newaxis] # (128, 125, 1)
+        phi = phi[:, :, tf.newaxis]
+
+        points = tf.concat([eta, phi], -1) # (128, 125, 2)
+
+        if masking:
+            reduced_fts = tf.expand_dims(tf.reduce_sum(features,-1),-1)  # From (128, 125, 22) to (128, 125, 1)
+            mask = tf.cast(tf.not_equal(reduced_fts, 0), dtype='float32')
```
Comment (ofivite) [inline models/particle_net.py]: To avoid possible scenarios with masking tokens for which sum of features is equal to 0 (probably even not realistic but still), I'd suggest to do compute mask as it is done currently in Transformer:
```python
mask = tf.math.reduce_any(tf.math.not_equal(features, 0), axis=-1)
mask = tf.cast(mask[:, :, tf.newaxis], dtype='float32')
```

File: configs/model/particle_net.yaml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+type: particle_net
+name: particle_net_test_vs_${vs_type}
+kwargs:
+  encoder:
+    conv_params: [
+      [16, [64, 64, 64]],
+      [16, [128, 128, 128]],
+      [16, [256, 256, 256]]
+    ]
+    dense_params: [256, 256]
```
Comment (ofivite) [inline configs/model/particle_net.yaml:22]: move `dense_params` to `kwargs.decoder` + changing it accordingly in the initialisation of `self.fc_params`

File: models/particle_net.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        self.edge_conv_layers = []
+        for layer_idx, layer_param in enumerate(self.conv_params):
+            K, channels = layer_param
+            self.edge_conv_layers.append(
+                EdgeConv(self.num_points, K, channels, with_bn=True, activation='relu', pooling=self.conv_pooling, name=f'{self.name}_EdgeConv_{layer_idx}')
+            )
+
+        self.dense_layers = []
+        self.dense_dropout = []
+        if self.fc_params is not None:
```
Comment (ofivite) [inline models/particle_net.py:50]: I suggest to wrap this set of dense and dropout layers together into `self.decoder_layers = tf.keras.Sequential(...)`, so that then in L77-81 it can be easily called with just line without a for loop:
```python
out = self.out(self.decoder_layers(x))
```

File: models/particle_net.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+        for idx, channel in enumerate(self.channels):
+            x = self.conv2d_layers[idx](x)
+            if self.with_bn:
+                x = self.batchnorm_layers[idx](x)
+            if self.activation:
+                x = self.activation_layers[idx](x)
+
+        if self.pooling == 'max':
+            fts = tf.reduce_max(x, axis=2)  # (N, P, C')
+        else:
```
Comment (ofivite) [inline models/particle_net.py]: add here the handling of 'average' case, and raise exception of other pooling is specified:
 ```python
if self.pooling == 'max':
    fts = tf.reduce_max(x, axis=2) # (N, P, C')
elif self.pooling == 'average':
    fts = tf.reduce_mean(x, axis=2) # (N, P, C')
else:
    raise RuntimeError('pooling parameter should be either max or mean')

File: models/particle_net.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+                        Dropout(dropout_rate)
+                    )
+            
+            self.out = Dense(self.num_classes, activation='softmax')
+
+    @tf.function
+    def call(self, input):
+        masking = True
+
+        features = input[0][:,:self.num_points,:].to_tensor() # (batch_size, particles, features), here: (128, 125, 22)
```
Comment (ofivite) [inline models/particle_net.py]: I think in theory it should be possible to drop the cut off (so usage of `num_points`) and then do knn for the whole set of (padded) constituents in the batch. (at least it worked for me in the jupyter setup)

Could you try removing `[:,:self.num_points,:]` here and then in `knn()` function retrieve the tensor shapes within the function itself, instead of passing `num_points` as the argument?

File: models/particle_net.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+            units, drop_rate = layer_param
+            x = self.dense_layers[layer_idx](x)
+            if drop_rate is not None and drop_rate > 0:
+                x = self.dense_dropout[layer_idx](x)
+
+        out = self.out(x)
+
+        return out  # (N, num_classes)
+
+    def create_padding_mask(self, seq):
```
Comment (ofivite) [inline models/particle_net.py]: remove unused function definition

**Review by ofivite — CHANGES_REQUESTED** [review]

File: models/particle_net.py (hunk 3 above)
Comment (ccmfb) [reply models/particle_net.py]: I don't know if I quite get what you mean. Do you mean that one can choose which type of coordinate system they want to use, i.e. choose between eta-phi or r-theta within the yaml?
If so, the distance function needs to be updated as it currently only works in eta-phi space.

File: models/particle_net.py (hunk 5 above)
Comment (ccmfb) [reply models/particle_net.py]: No, I think you are right. I missed that.

File: configs/model/particle_net.yaml (hunk 7 above)
Comment (ccmfb) [reply configs/model/particle_net.yaml:22]: Yes, I suppose I should also move `dropout_rate` as it is only used with the `dense_params`?

File: models/particle_net.py (hunk 10 above)
Comment (ccmfb) [reply models/particle_net.py]: The issue with retrieving the shape seems to be that during the execution tf cannot access the shape without evaluating the whole tensor. You can force tf to always evaluate them (with `run_eagerly=True` as an argument in the model.compile() function), but this significantly increases the training time.
I am not sure how to deal with this issue.

File: models/particle_net.py (hunk 4 above)
Comment (ccmfb) [reply models/particle_net.py:91]: I presume so, it is from this implementation [https://github.com/hqucms/ParticleNet/blob/master/tf-keras/tf_keras_model.py](url) in line 93.

File: models/particle_net.py [hunk 12]
```diff
@@ -1,17 +1,20 @@
+from tkinter import W
```
Comment (ofivite) [inline models/particle_net.py]: remove unused import

File: models/particle_net.py [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+            self.decoder_layers = tf.keras.Sequential()
+
+            for layer_idx, layer_param in enumerate(self.fc_params):
+                units, dropout_rate = layer_param
+
+                self.decoder_layers.add(Dense(units, activation='relu'))                
+                if dropout_rate is not None and dropout_rate > 0:
+                    self.decoder_layers.add(Dropout(dropout_rate))
+
+            self.decoder_layers.add(Dense(self.num_classes, activation='softmax'))
```
Comment (ofivite) [inline models/particle_net.py]: tabulate it out of the `if` statement, so that this layer is always added:
```suggestion
        self.decoder_layers.add(Dense(self.num_classes, activation='softmax'))
```

File: models/particle_net.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+        
+        self.edge_conv_layers = []
+        for layer_idx, layer_param in enumerate(self.conv_params):
+            K, channels = layer_param
+            self.edge_conv_layers.append(
+                EdgeConv(self.num_points, K, channels, with_bn=True, activation='relu', pooling=self.conv_pooling, name=f'{self.name}_EdgeConv_{layer_idx}')
+            )
+
+        if self.fc_params is not None:
+            self.decoder_layers = tf.keras.Sequential()
```
Comment (ofivite) [inline models/particle_net.py]: move `self.decoder_layers` init one line up out of the `if` scope, so that it is always initialised and filled at least with `Dense(self.num_classes, activation='softmax')` (see also comment below)

File: models/particle_net.py [hunk 15]
```diff
@@ -140,8 +125,10 @@ def call(self, points, features):
 
         if self.pooling == 'max':
             fts = tf.reduce_max(x, axis=2)  # (N, P, C')
-        else:
+        elif self.pooling == 'mean':
```
Comment (ofivite) [inline models/particle_net.py:159]: if using "mean" identifier for this operation, please also change everywhere `'average'->'mean'` (default params of `EdgeConv` init and in `particle_net.yaml`)

File: configs/model/particle_net.yaml (hunk 7 above)
Comment (ofivite) [reply configs/model/particle_net.yaml:22]: indeed, implementation in the new commit looks good 👍🏻

File: models/particle_net.py (hunk 10 above)
Comment (ofivite) [reply models/particle_net.py]: Right, so as we've checked, it seems to be possible to remove `num_points ` here and all other places if accessing the shape of the tensor within `knn()` function, opposite to passing it as the first argument:
```python
# inside def knn()
# ...
queries_shape = tf.shape(features)
batch_size = queries_shape[0]
num_points = queries_shape[1]
```

File: models/particle_net.py (hunk 3 above)
Comment (ofivite) [reply models/particle_net.py]: Yeah that would be nice to have it, but the way distance is computed is already hardwired into the model and it's hard to make it flexible straight away.

So my idea was to add a switch between using polar coordinates `["r", "theta"]` and user-defined set `["a", "b", "c", ...]`. It can be called `coordinate_type` and if it is `"polar"`, then there is assertion to be added `assert spatial_features == ["r", "theta"]` followed by current L46-52 without change. 

Else, it converts `spatial_features` to a list of indices, which are then gathered from the inputs into `points` similarly to how it is done in the embedding:
https://github.com/ofivite/tatoo/blob/6c1e019bb77f4fb2a0427a26e71c40a84603724b/models/embedding.py#L20

File: models/particle_net.py [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+
+                self.decoder_layers.add(Dense(units, activation='relu'))                
+                if dropout_rate is not None and dropout_rate > 0:
+                    self.decoder_layers.add(Dropout(dropout_rate))
+
+            self.decoder_layers.add(Dense(self.num_classes, activation='softmax'))
+
+    @tf.function
+    def call(self, input):
+        features = input[0][:,:self.num_points,:].to_tensor() # (batch_size, particles, features), here: (128, 125, 22)
```
Comment (ofivite) [inline models/particle_net.py]: As discussed, please add here an embedding of inputs to enable dropping functionality for ParticleNet as well, following exactly the approach of how it's done currently in Transformer:
https://github.com/ofivite/tatoo/blob/6c1e019bb77f4fb2a0427a26e71c40a84603724b/models/transformer.py#L127-L143

**Review by ofivite — CHANGES_REQUESTED** [review]

**Review by ofivite — APPROVED** [review]
looks good to me, thanks for the integration @ccmfb !
